# Pré-processamento de dados

## Contexto
Limpeza, transformação, conversão e manipulação de um dataset 'cru', para um dataset pronto para ser treinado para machine learning.

## 1. Carregamento e análise inicial dos dados

Carregando dataset e visualizando informações básicas para entender sua estrutura.

In [313]:
import pandas as pd

df = pd.read_excel('base_aed_evasao_pre_processamento.xlsx')

In [314]:
df.head()

,Idade,Gênero,Estado Civil,Renda Familiar (R$),Escola Ensino Médio,Status,Motivo da Evasão,Trabalha Atualmente,Último Período Cursado,Desempenho,Renda Familiar Total
0,18,feminino,Solteiro(a),1414.37,Pública,Ativo,NaN,Não,5,68.5,1414.37
1,20,FEM,Divorciado(a),4238.37,Privada,Evadido,Dificuldade financeira,Não,6,76.4,4238.37
2,18,Masculino,Viúvo(a),2295.80,Pública,Ativo,NaN,Não,9,77.8,2295.80
3,27,masc,Solteiro(a),3765.94,Pública,Evadido,Baixo desempenho,Sim,1,61.9,3765.94
4,30,Masculino,Solteir,3991.39,Privada,Ativo,NaN,Sim,1,61.6,3991.39


In [315]:
print(df.columns)
print()
print(df.info())
print()
df.describe()

Index(['Idade', 'Gênero', 'Estado Civil', 'Renda Familiar (R$)',
       'Escola Ensino Médio', 'Status', 'Motivo da Evasão',
       'Trabalha Atualmente', 'Último Período Cursado', 'Desempenho',
       'Renda Familiar Total'],
      dtype='object')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Idade                   153 non-null    int64  
 1   Gênero                  153 non-null    object 
 2   Estado Civil            153 non-null    object 
 3   Renda Familiar (R$)     145 non-null    float64
 4   Escola Ensino Médio     145 non-null    object 
 5   Status                  153 non-null    object 
 6   Motivo da Evasão        51 non-null     object 
 7   Trabalha Atualmente     145 non-null    object 
 8   Último Período Cursado  153 non-null    int64  
 9   Desempenho              145 non-null    float64
 10  Rend

,Idade,Renda Familiar (R$),Último Período Cursado,Desempenho,Renda Familiar Total
count,153.000000,145.000000,153.000000,145.000000,145.000000
mean,24.444444,2888.185517,5.562092,73.206207,2888.185517
std,5.185873,4076.689059,2.939813,9.724570,4076.689059
min,17.000000,500.000000,1.000000,61.500000,500.000000
25%,20.000000,1814.970000,3.000000,67.000000,1814.970000
50%,24.000000,2471.130000,6.000000,72.200000,2471.130000
75%,29.000000,3268.130000,8.000000,78.400000,3268.130000
max,35.000000,50000.000000,10.000000,150.000000,50000.000000


### 1.1 Valores nulos e coluna duplicada

A partir dessa análise inicial, notamos que:
1. a coluna 'Renda Familiar Total' é duplicada em relação à coluna 'Renda Familiar (R$)';
2. as colunas 'Renda Familiar (R$)', 'Escola Ensino Médio', 'Motivo da Evasão', 'Trabalha Atualmente' e 'Desempenho' possuem valores nulos - Tendo em vista que o dataframe possue 153 linhas;
3. as colunas 'Gênero' e 'Estado Civil' possuem valores categóricos não padronizadods. Ex.: 'feminino' e 'FEM', 'Solteiro(a)' e 'Solteir';
4. existem valores outliers nas colunas 'Renda Familiar (R$)' e 'Desempenho'.

### 1.2 Criando dataframe filtrado

Filtrando dataframe apenas para as colunas que iremos utilizar, removendo a coluna duplicada (Renda Familiar Total).

In [316]:
df = df[['Idade', 'Gênero', 'Estado Civil', 'Renda Familiar (R$)', 'Escola Ensino Médio', 'Status', 'Motivo da Evasão', 'Trabalha Atualmente', 'Último Período Cursado', 'Desempenho']].copy()

## 2. Pré-processamento

### 2.1 Limpeza de dados

#### 2.1.1 Registros/linhas duplicados

##### 2.1.1.1 Verificando se há linhas duplicadas no DF

In [317]:
duplicados = df.duplicated(keep=False)

print(f'Linhas duplicadas: {duplicados.sum()}')

df[duplicados]

Linhas duplicadas: 6


,Idade,Gênero,Estado Civil,Renda Familiar (R$),Escola Ensino Médio,Status,Motivo da Evasão,Trabalha Atualmente,Último Período Cursado,Desempenho
14,27,Masculino,Viúvo(a),4892.37,Pública,Ativo,NaN,Sim,6,69.4
75,17,Feminino,Viúvo(a),899.56,Privada,Evadido,Baixo desempenho,Não,6,77.6
98,20,Feminino,Divorciado(a),2912.31,Pública,Evadido,Mudança de curso,Sim,6,71.2
150,27,Masculino,Viúvo(a),4892.37,Pública,Ativo,NaN,Sim,6,69.4
151,20,Feminino,Divorciado(a),2912.31,Pública,Evadido,Mudança de curso,Sim,6,71.2
152,17,Feminino,Viúvo(a),899.56,Privada,Evadido,Baixo desempenho,Não,6,77.6


Notamos repetição entre as linhas dos índices abaixo:
- 14 e 150
- 75 e 152
- 98 e 151

Assim, será necessário remover 3 linhas, evitando a duplicidade.

##### 2.1.1.2 Excluindo registros duplicados e verificando sucesso do ajuste: 

In [318]:
df = df.drop_duplicates().copy()
print(f'Linhas duplicadas: {df.duplicated().sum()}')

Linhas duplicadas: 0


#### 2.1.2 Valores nulos:

Nas seções 1 e 1.1 notamos que o dataframe possui algumas colunas com valores nulos. Vamos investigar esses valores nulos mais a fundo. 

In [319]:
print(f'Shape do DF sem duplicadas: {df.shape[0]}')

nulos = df.isna().sum()
print(f'\nQuantidade de valores nulos em cada coluna:\n{nulos}')

Shape do DF sem duplicadas: 150

Quantidade de valores nulos em cada coluna:
Idade                       0
Gênero                      0
Estado Civil                0
Renda Familiar (R$)         8
Escola Ensino Médio         8
Status                      0
Motivo da Evasão          101
Trabalha Atualmente         8
Último Período Cursado      0
Desempenho                  8
dtype: int64


* A coluna 'Motivo da Evasão' apresenta 101 valores nulos (alunos que estão ativos ou se formaram), ou seja, 49 alunos que evadiram (150 - 101 = 49).

* Notamos que as colunas 'Renda Familiar ($)', 'Escola Ensino Médio', 'Trabalha Atualmente' e 'Desempenho' possuem a mesma de quantidade de valores nulos: 8.

##### 2.1.2.1 Validando evadidos com motivo de evasão nulo

Para validar se esse número realmente representa a quantidade de alunos evadidos e garantir que não haja valores não preenchidos, realizamos uma verificação da quantidade de alunos evadidos baseado no shape do DF:

In [320]:
evasao = df[df['Status'] == 'Evadido']
print(f'Quantidade de alunos evadidos: {evasao.shape[0]}')

Quantidade de alunos evadidos: 54


* Notamos uma discrepância entre os valores, onde anteriormente tínhamos o valor de 49 alunos evadidos, agora percebemos que há na verdade 54 alunos evadidos.

* Isso indica que há 5 alunos evadidos com a coluna 'Motivo da Evasão' não preenchida.

Vamos validar filtrando pelos alunos evadidos e com a coluna em questão com valores nulos:

In [321]:
evadidos_sem_motivo = df[(df['Status'] == 'Evadido') & (df['Motivo da Evasão'].isna())]
print(f'Quantidade de registros evadidos sem motivo de evasão (nulo): {evadidos_sem_motivo.shape[0]}')

Quantidade de registros evadidos sem motivo de evasão (nulo): 5


##### 2.1.2.2 Substituindo valores nulos de evadidos sem motivo

Vamos substituir esses valores nulos por 'Não Informado' para não introduzir viés em nosso DF, e em seguida validar se os valores nulos para o status evadido foram zerados.

In [322]:
# Substituindo valores nulos por 'Não Informado'
filtro = (df['Status'] == 'Evadido') & (df['Motivo da Evasão'].isna())
df.loc[filtro, 'Motivo da Evasão'] = 'Não Informado'

# Verificando ajuste
evadidos_sem_motivo = df[(df['Status'] == 'Evadido') & (df['Motivo da Evasão'].isna())]
print(f'Quantidade de registros evadidos sem motivo de evasão (nulo): {evadidos_sem_motivo.shape[0]}')

Quantidade de registros evadidos sem motivo de evasão (nulo): 0


##### 2.1.2.3 Substituindo valores nulos da coluna 'Desempenho'

Antes de preencher os valores nulos da coluna 'Desempenho', vamos calcular o valor da mediana sem considerar valores maiores que 100, transformar esse valor em nulo, em seguida imputar o valor da mediana nessa posição, e então validar se está devidamente ajustado.

In [323]:
# Transformando valor de 150 em nulo 
df.loc[df['Desempenho'] > 100, 'Desempenho'] = None

# Definindo mediana e substituindo no df
mediana_desempenho = df['Desempenho'].median()
df['Desempenho'] = df['Desempenho'].fillna(mediana_desempenho)

df['Desempenho'].describe()

count    150.000000
mean      72.643333
std        7.162621
min       61.500000
25%       67.025000
50%       72.200000
75%       77.800000
max       88.700000
Name: Desempenho, dtype: float64

##### 2.1.2.4 Substituindo valores nulos da coluna 'Renda Familiar (R$)'

Nesta coluna, vamos realizar o mesmo procedimento realizado na coluna Desempenho, tendo em vista o valor outlier de 50000.

Com isso, vamos calcular o valor da mediana sem considerar esse valor de 500000, transformar esse valor em nulo, em seguida imputar o valor da mediana nessa posição, e então validar se está devidamente ajustado.

In [324]:
# Visualizando as 5 maiores rendas
print(f'5 maiores rendas:\n{df['Renda Familiar (R$)'].sort_values(ascending=False).head()}')


# Transformando valor de 50000 em nulo 
df.loc[df['Renda Familiar (R$)'] > 4900, 'Renda Familiar (R$)'] = None


# Definindo mediana e substituindo no df
mediana_renda = df['Renda Familiar (R$)'].median()
df['Renda Familiar (R$)'] = df['Renda Familiar (R$)'].fillna(mediana_renda)

print('\nEstatísticas após ajuste:')
df['Renda Familiar (R$)'].describe()

5 maiores rendas:
5     50000.00
14     4892.37
11     4861.60
82     4843.47
65     4700.70
Name: Renda Familiar (R$), dtype: float64

Estatísticas após ajuste:


count     150.000000
mean     2548.529667
std      1003.331879
min       500.000000
25%      1940.707500
50%      2466.310000
75%      3175.932500
max      4892.370000
Name: Renda Familiar (R$), dtype: float64

##### 2.1.2.5 Substituindo valores nulos da coluna 'Escola Ensino Médio'.

Nesta situação, para sermos conservadores, vamos alterar os valores nulos para "Desconhecida" e verificar a quantidade de valores nulos para validar se o ajuste foi realizado com sucesso.

In [325]:
df['Escola Ensino Médio'] = df['Escola Ensino Médio'].fillna('Desconhecida')

print(df['Escola Ensino Médio'].isna().sum())

0


##### 2.1.2.6 Substituindo valores nulos da coluna 'Trabalha Atualmente'

Vamos realizar o mesmo processo da coluna 'Escola Ensino Médio', imputando 'Desconhecido' aos valores nulos.

In [326]:
df['Trabalha Atualmente'] = df['Trabalha Atualmente'].fillna('Desconhecido')

print(df['Trabalha Atualmente'].isna().sum())

0


##### 2.1.2.7 Validação de ajustes
Podemos observar que agora não temos mais valores nulos (desconsiderando os 96 registros nulos de Motivos da Evasão, que representam alunos ativos ou formados):

In [327]:
df.isna().sum()

Idade                      0
Gênero                     0
Estado Civil               0
Renda Familiar (R$)        0
Escola Ensino Médio        0
Status                     0
Motivo da Evasão          96
Trabalha Atualmente        0
Último Período Cursado     0
Desempenho                 0
dtype: int64

### 2.2 Transformação de dados
#### 2.2.1 Padronizando valores de texto
Ao visualizar as primeiras linhas do DF, podemos observar uma inconsistência no padrão dos valores da coluna 'Gênero':

In [328]:
df.head()

,Idade,Gênero,Estado Civil,Renda Familiar (R$),Escola Ensino Médio,Status,Motivo da Evasão,Trabalha Atualmente,Último Período Cursado,Desempenho
0,18,feminino,Solteiro(a),1414.37,Pública,Ativo,NaN,Não,5,68.5
1,20,FEM,Divorciado(a),4238.37,Privada,Evadido,Dificuldade financeira,Não,6,76.4
2,18,Masculino,Viúvo(a),2295.80,Pública,Ativo,NaN,Não,9,77.8
3,27,masc,Solteiro(a),3765.94,Pública,Evadido,Baixo desempenho,Sim,1,61.9
4,30,Masculino,Solteir,3991.39,Privada,Ativo,NaN,Sim,1,61.6


Vamos validar em quais colunas a mesma inconsistência ocorre:

Notamos que as colunas 'Gênero' e 'Estado Civil' possuem mais de uma palavra para um mesmo dado. Com isso, vamos padronizar e organizar estes dados, validando se os ajustes ocorreram corretamente.

In [329]:
for col in df.columns:
    print(df[col].value_counts())
    print()

Idade
21    13
20    12
18    12
24    12
17    11
25    10
33    10
29     9
27     8
23     8
30     8
22     8
19     6
28     5
34     5
26     4
32     4
31     4
35     1
Name: count, dtype: int64

Gênero
Feminino     57
Masculino    47
Outro        43
feminino      1
FEM           1
masc          1
Name: count, dtype: int64

Estado Civil
Casado(a)        46
Viúvo(a)         39
Divorciado(a)    34
Solteiro(a)      27
Solteir           1
Casada            1
Viuvo             1
Separado          1
Name: count, dtype: int64

Renda Familiar (R$)
2466.31    10
500.00      2
2295.80     1
1414.37     1
3765.94     1
           ..
4169.38     1
2966.37     1
2471.13     1
3141.08     1
722.38      1
Name: count, Length: 140, dtype: int64

Escola Ensino Médio
Privada         78
Pública         64
Desconhecida     8
Name: count, dtype: int64

Status
Ativo      65
Evadido    54
Formado    31
Name: count, dtype: int64

Motivo da Evasão
Baixo desempenho          18
Dificuldade financeira    

In [330]:
#Gênero
#Feminino     57
#Masculino    47
#Outro        43
#feminino      1
#FEM           1
#masc          1

df['Gênero'] = df['Gênero'].replace(['feminino', 'FEM'], 'Feminino')
df['Gênero'] = df['Gênero'].replace('masc', 'Masculino')

#Estado Civil
#Casado(a)        46
#Viúvo(a)         39
#Divorciado(a)    34
#Solteiro(a)      27
#Solteir           1
#Casada            1
#Viuvo             1
#Separado          1

df['Estado Civil'] = df['Estado Civil'].replace('Solteir', 'Solteiro(a)')
df['Estado Civil'] = df['Estado Civil'].replace('Casada', 'Casado(a)')
df['Estado Civil'] = df['Estado Civil'].replace('Viuvo', 'Viúvo(a)')
df['Estado Civil'] = df['Estado Civil'].replace('Separado', 'Divorciado(a)')


print(df['Gênero'].value_counts())
print()
print(df['Estado Civil'].value_counts())

Gênero
Feminino     59
Masculino    48
Outro        43
Name: count, dtype: int64

Estado Civil
Casado(a)        47
Viúvo(a)         40
Divorciado(a)    35
Solteiro(a)      28
Name: count, dtype: int64


#### 2.2.2 Criação de variáveis derivadas

In [331]:
# Variáveis categóricas para idade:
df['Faixa Etária'] = pd.cut(df['Idade'],
                            bins = [0, 21, 49, 999],
                            labels = ['Jovem', 'Adulto', 'Maduro'])

# Normalizando renda:
min_renda = df['Renda Familiar (R$)'].min()
max_renda = df['Renda Familiar (R$)'].max()

df['Renda Familiar Normalizada'] = (df['Renda Familiar (R$)'] - min_renda) / (max_renda - min_renda)


df.head()

,Idade,Gênero,Estado Civil,Renda Familiar (R$),Escola Ensino Médio,Status,Motivo da Evasão,Trabalha Atualmente,Último Período Cursado,Desempenho,Faixa Etária,Renda Familiar Normalizada
0,18,Feminino,Solteiro(a),1414.37,Pública,Ativo,NaN,Não,5,68.5,Jovem,0.208172
1,20,Feminino,Divorciado(a),4238.37,Privada,Evadido,Dificuldade financeira,Não,6,76.4,Jovem,0.851105
2,18,Masculino,Viúvo(a),2295.80,Pública,Ativo,NaN,Não,9,77.8,Jovem,0.408845
3,27,Masculino,Solteiro(a),3765.94,Pública,Evadido,Baixo desempenho,Sim,1,61.9,Adulto,0.743548
4,30,Masculino,Solteiro(a),3991.39,Privada,Ativo,NaN,Sim,1,61.6,Adulto,0.794876


#### 2.2.3 Conversão de variáveis categóricas para números
Aplicamos 3 estratégias distintas conforme a natureza de cada variável:

- **Variável target**: `Status` convertida para binário (1 = Evadido, 0 = demais);
- **Mapeamento ordinal** para variáveis com ordem natural (Faixa Etária);
- **One-Hot Encoding (OHE)** para variáveis sem hierarquia (Gênero, 
Estado Civil, Escola Ensino Médio, Status, Trabalha Atualmente), 
evitando que o modelo interprete uma ordem inexistente entre as categorias.

In [332]:
# Mapeamento ordinal (variáveis com ordem natural)
mapeamento_ordinal = {'Faixa Etária': {'Jovem': 1, 'Adulto': 2, 'Maduro': 3}}


df_convertido = df[['Idade', 'Gênero', 'Estado Civil', 'Renda Familiar (R$)', 'Escola Ensino Médio', 'Trabalha Atualmente', 'Último Período Cursado', 'Desempenho', 'Faixa Etária']].copy()

for col, mapa in mapeamento_ordinal.items():
    df_convertido[col] = df_convertido[col].map(mapa)


# One-Hot Encoding (variáveis sem ordem natural)
colunas_ohe = ['Gênero', 'Estado Civil', 'Escola Ensino Médio', 'Trabalha Atualmente']
df_convertido = pd.get_dummies(df_convertido, columns = colunas_ohe, dtype = int)


# Definindo variável target (o que queremos prever)
df_convertido['Target'] = (df['Status'] == 'Evadido').astype(int)

df_convertido.head()

,Idade,Renda Familiar (R$),Último Período Cursado,Desempenho,Faixa Etária,Gênero_Feminino,Gênero_Masculino,Gênero_Outro,Estado Civil_Casado(a),Estado Civil_Divorciado(a),Estado Civil_Solteiro(a),Estado Civil_Viúvo(a),Escola Ensino Médio_Desconhecida,Escola Ensino Médio_Privada,Escola Ensino Médio_Pública,Trabalha Atualmente_Desconhecido,Trabalha Atualmente_Não,Trabalha Atualmente_Sim,Target
0,18,1414.37,5,68.5,1,1,0,0,0,0,1,0,0,0,1,0,1,0,0
1,20,4238.37,6,76.4,1,1,0,0,0,1,0,0,0,1,0,0,1,0,1
2,18,2295.80,9,77.8,1,0,1,0,0,0,0,1,0,0,1,0,1,0,0
3,27,3765.94,1,61.9,2,0,1,0,0,0,1,0,0,0,1,0,0,1,1
4,30,3991.39,1,61.6,2,0,1,0,0,0,1,0,0,1,0,0,0,1,0


## 3. Exportação do DataFrame como CSV

In [333]:
df.to_csv('base_aed_evasao_pre_processamento_tratada.csv', index=False)

## Conclusão

Dataset processado e exportado com as seguintes alterações:
- Coluna duplicada removida;
- 3 linhas duplicadas removidas;
- Valores nulos tratados adequadamente de acordo com cada coluna;
- Valores outliers das colunas 'Desempenho' e 'Renda Familiar ($)' substituidos pela mediana;
- Padronização das variáveis categóricas das colunas 'Gênero' e 'Estado Civil';
- Criação de colunas 'Faixa Etária' e 'Renda Familiar Normalizada' categóricas derivadas de 'Idade' e 'Renda Familiar (R$)' respectivamente;
- Variável target e encoding, com mapeamento ordinal e One-Hot Encoding;
- Exportação do dataframe para planilha em formato csv.